state-chain/pallets/cf-elections/src/electoral_systems/solana_swap_accounts_tracking.rs

```diff
@@ -0,0 +1,277 @@
+		_previous_vote: Option<(VotePropertiesOf<Self>, AuthorityVoteOf<Self>)>,
+		_vote: &<Self::Vote as VoteStorage>::PartialVote,
+	) -> Result<VotePropertiesOf<Self>, CorruptStorageError> {
+		Ok(())
+	}
+
+	fn is_vote_desired<ElectionAccess: ElectionReadAccess<ElectoralSystem = Self>>(
+		_election_identifier_with_extra: crate::electoral_system::ElectionIdentifierOf<Self>,
+		_election_access: &ElectionAccess,
+		_current_vote: Option<(VotePropertiesOf<Self>, AuthorityVoteOf<Self>)>,
+	) -> Result<bool, CorruptStorageError> {
+		Ok(true)
+	}
+
+	fn is_vote_needed(
+		(_, current_partial_vote, _): (
+			VotePropertiesOf<Self>,
+			<Self::Vote as VoteStorage>::PartialVote,
+			AuthorityVoteOf<Self>,
+		),
+		(proposed_partial_vote, _): (
+			<Self::Vote as VoteStorage>::PartialVote,
+			<Self::Vote as VoteStorage>::Vote,
+		),
+	) -> bool {
+		current_partial_vote != proposed_partial_vote
+	}
+
+	fn on_finalize<ElectoralAccess: ElectoralWriteAccess<ElectoralSystem = Self>>(
+		electoral_access: &mut ElectoralAccess,
+		election_identifiers: Vec<ElectionIdentifier<Self::ElectionIdentifierExtra>>,
+		current_block_number: &Self::OnFinalizeContext,
+	) -> Result<Self::OnFinalizeReturn, CorruptStorageError> {
+		if let Some(election_identifier) = election_identifiers
+			.into_iter()
+			.at_most_one()
+			.map_err(|_| CorruptStorageError::new())?
+		{
+			let mut election_access = electoral_access.election_mut(election_identifier)?;
+			if let Some(consensus) = election_access.check_consensus()?.has_consensus() {
+				let mut known_accounts = election_access.properties()?;
+				election_access.delete();
+				known_accounts.witnessed_open_accounts.extend(consensus.new_accounts.iter().map(
+					|(account, swap_details)| {
+						Hook::initiate_vault_swap((*swap_details).clone());
+						(*account).clone()
+					},
+				));
+				consensus.confirm_closed_accounts.into_iter().for_each(|acc| {
+					known_accounts.closure_initiated_accounts.remove(&acc);
+				});
+
+				if Hook::get_number_of_available_sol_nonce_accounts() >
+					NONCE_AVAILABILITY_THRESHOLD_FOR_INITIATING_SWAP_ACCOUNT_CLOSURES &&
+					(known_accounts.witnessed_open_accounts.len() >=
+						MAX_BATCH_SIZE_OF_VAULT_SWAP_ACCOUNT_CLOSURES ||
+						(*current_block_number)
+							.checked_sub(&electoral_access.unsynchronised_state()?)
+							.expect("current block number is always greater than when apicall was last created")
+							.into() >= MAX_WAIT_BLOCKS_FOR_SWAP_ACCOUNT_CLOSURE_APICALLS)
+				{
+					let accounts_to_close: Vec<_> = if known_accounts.witnessed_open_accounts.len() >
+						MAX_BATCH_SIZE_OF_VAULT_SWAP_ACCOUNT_CLOSURES
+					{
+						known_accounts
+							.witnessed_open_accounts
+							.drain(..MAX_BATCH_SIZE_OF_VAULT_SWAP_ACCOUNT_CLOSURES)
+							.collect()
+					} else {
+						sp_std::mem::take(&mut known_accounts.witnessed_open_accounts)
+					};
+					match Hook::close_accounts(accounts_to_close.clone()) {
```

Comment:
depending on whether we were successfully able to close accounts, we have to set the unsynchronized state so it makes sense to have this logic here.

Reply:
If all the logic were in the close_accounts hook, then, we wouldn't need to have two separate hooks for `get_number_of_available_sol_nonce_accounts` and `close_accounts` - close accounts would already have access to it -> which makes it easier to mock and test, and for the writing of state, all you need is to return a result as is already the case, and just set the block number.

Then for testing this ElectoralSystem's on_finalize, it's very simple, you can just check "has hook been called" - we already have tests that do this for other ElectoralSystems so it would be very similar. Then you could also write tests for the close_accounts hook implementation itself, separately.

Reply:
I considered moving all the logic concerning all three hooks inside one hook but realized it would not actually make testing easier. This is because in that case, writing the test for one big hook impl would then be difficult since the logic in hook impl calls external code (constructing apicall, accessing sol nonces etc). So we would have to mock those. So basically we are back where we started. We wanted to avoid create 3 mocks of 3 election hooks, instead we create one. However, that means we create those extra mocks while testing the hook impl so I dont think we gain anything. In fact, creating the mocks for the hook impl is even harder since we dont have a framework for it at the Runtime level. creating 3 mocks instead of one while writing tests for the on_finalize shouldnt be a big deal.

Reply:
Having the apicall construction be a separate trait function is probably reasonable, it's also kind of a different part of the flow, so conceptually could make things easier to follow. Though I think putting the get nonce and close accounts in one should definitely be simpler

Reply:
I am not sure I understand what you mean. The apicall construction is inside the ```close_accounts``` function. trying to combine apicall construction and get nonces into one hook would mean we have to move the whole logic inside that hook which creates the problem of testability that I mentioned above.

In general, I think it is a pretty arbitrary and subjective choice here between having the logic in one place or splitting it. In the end logic is logic, one has to follow it to make sure it works. Its not like we don't have long pieces of logic inside single functions in other places in the codebase.

Reply:
Sorry, should've been clearer. I mean leaving initiate_vault_swap separate makes sense. But since the only reason we need the number of sol accounts is to decide whether we construct the close accounts call, seems to make sense to include it.

I'm not against having long pieces of logic, it's just about keeping conceptually similar things together and the boundaries between things clean. Happy to see how the tests look keeping it like this

Reply:
yes of course but in this case we cant test the hook impl easily as i said before so it makes sense to keep it here  since we want to test and keep the hook impl trivial.